intermediate_source/parametrizations.py

```diff
@@ -0,0 +1,353 @@
+print(layer.parametrizations.weight.original)
+
+###############################################################################
+# Besides these two small differences, the parametrization is doing exactly the same
+# as our manual implementation
+symmetric = Symmetric()
+assert torch.allclose(layer.weight, symmetric(layer.parametrizations.weight.original))
+
+###############################################################################
+# Parametrizations are first-class citizens
+# -----------------------------------------
+# Since ``layer.parametrizations`` is an `nn.ModuleList`, it means that the parametrizations
+# are properly registered as submodules of the original module. As such, the same rules
+# for registering parameters in a module apply to register a parametrization.
+# For example, if a parametrization has parameters, these will be moved from CPU
+# to CUDA when calling ``model = model.cuda()``.
+#
+# Caching the value of a parametrization
+# --------------------------------------
+# Parametrizations come with an inbuilt caching system via the context manager ``P.cached()``
+class NoisyParametrization(nn.Module):
+    def forward(self, X):
+        print("Computing the Parametrization")
+        return X
+
+layer = nn.Linear(2, 3)
+P.register_parametrization(layer, "weight", NoisyParametrization())
+print("Here, layer.weight is recomputed every time we call it")
+foo = layer.weight + layer.weight.T
+bar = layer.weight.sum()
+with P.cached():
+    print("Here, it is computed just the first time layer.weight is called")
+    foo = layer.weight + layer.weight.T
+    bar = layer.weight.sum()
+
+###############################################################################
+# Composing Parametrizations
+# --------------------------
+# Concatenating two parametrizations is as easy as registering them on the same tensor.
+# We may use this to create complex parametrizations from simple ones. For example, the
+# `Cayley map <https://en.wikipedia.org/wiki/Cayley_transform#Matrix_map>`_
+# maps the skew-symmetric matrices to the orthogonal matrices of positive determinant. We can
+# concatenate ``Skew`` and a parametrization that implements the Cayley map to get a layer with
+# orthogonal weight
+class CayleyMap(nn.Module):
+    def __init__(self, n):
+        self.register_buffer("Id", torch.eye(n))
+
+    def forward(self, X):
+        # (I + X)(I - X)^{-1}
+        return torch.solve(self.Id + X, self.Id - X).solution
+
+layer = nn.Linear(3, 3)
+P.register_parametrization(layer, "weight", Skew())
+P.register_parametrization(layer, "weight", CayleyMap(3))
+X = layer.weight
+assert torch.allclose(X.T @ X, torch.eye(3))  # X is orthogonal
+
+###############################################################################
+# This may also be used to prune a parametrized module, or to reuse parametrizations. For example,
+# we may use the fact that the exponential of matrices maps the symmetric matrices to the
```

Comment:
While this sentence makes sense if you read it slowly, I feel like we could make it nicer.